### PR TITLE
kv: opt into Elastic CPU limiting resume spans

### LINKED ISSUE #97886: kvserver: fix incorrect handling of ExportRequests that are pre-empted by the CPU limiter
In https://github.com/cockroachdb/cockroach/pull/96691 we enabled the elastic CPU limiter for all users of ExportRequests. Since then we have thought of two bugs that were an oversight:

1. As surfaced by https://github.com/cockroachdb/cockroach/issues/97693, if the client does not set a TargetBytes on the ExportRequest then DistSender may split the request if it targets >1 range. These requests are then sent in parallel and the responses are combined before returning to the client. The logic that combines the response headers does not expect both the LHS and RHS to have a resume span. With elastic CPU limiting enabled this is very much possible and results in the ExportRequest erroring out https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvpb/api.go#L323-L329.

2. In a mixed version cluster ExportRequest clients running on older nodes do not know to handle resume spans. Thus, if a span gets pre-empted by the elastic CPU limiter then the client will not resend the remaining span to be exported.

To address 2) we plan to add a bool to the `AdmissionHeader` that indicates whether or not the client is capable of handling paginated ExportRequests. This field will default to false if the client is on older nodes. If the field is true, we will return from kvserver to the client on exceeding CPULimit. If false, we will fall back to retrying the ExportRequest on the server side, as was the case before #96691.

To address 1), we will teach DistSender to not parallelize ExportRequests that are sent with the newly added `AdmissionHeader` bool set to true. This is a short-term fix but we expect the loss of parallelization to only affect `crdb_internal.fingerprint` since the other call sites are all on the descriptors table which is _usually_ on a single range. A longer-term fix will be to teach DistSender to combine the responses of parallel requests into a slice of resume spans and then have the client handle multiple, non-overlapping resume spans.

cc: @stevendanna 

Jira issue: CRDB-24946

Epic CRDB-21944

### PR DESCRIPTION
**Pull request by stevendanna**
In #96691, we changed ExportRequest such that it returns resume spans that result from the elastic CPU limiter all the way to the caller.

This has at least two problems:

1) In a mixed-version state, the caller might not yet know how to
   handle resume spans. This could result in incomplete responses
   erroneous being used as if they were full responses.

2) The DistSender inspects a request to determine whether it may stop
   early. If it shouldn't be able to stop early, then the request is
   split up, possibly sent in parallel, and all responses are
   combined. The code which combines responses asserts that neither side
   has a resume span.  As a result, we've seen failures such as

        crdb_internal.fingerprint(): combining
        /Tenant/2/Table/106/1/-{8403574544142222370/0-37656332809536692}
        with /Tenant/{2/Table/106/1/436440321206557763/0-3}

   since the change was made.

Here, we add a new request header field to allow callers to indicate whether they are prepared to accept resume spans. Further, we add that new field into the logic in DistSender which decides how to process requests.

The downside here is that crdb_internal.fingerprint won't have its requests sent in parallel.

Release note: None

Fixes #97886,  #97903

Epic: none

### CODE REVIEW

File: pkg/kv/kvpb/api.proto [hunk 1]
```diff
@@ -2635,6 +2635,15 @@ message Header {
   // each Scan and ReverseScan.
   sql.sqlbase.IndexFetchSpec index_fetch_spec = 29;
 
+  // ReturnElasticCPUResumeSpans, if set, indicates that the caller
+  // expects early-termination of requests based on the Elastic CPU
+  // limiter.
+  //
+  // Resume spans returned because of the unerlying request being
```
Comment (adityamaru) [inline pkg/kv/kvpb/api.proto:2642]: nit: s/unerlying/underlying/g

**Review by adityamaru — APPROVED** [review]
Thanks for this! For posterity noting that this boolean also ensures that in a mixed-version cluster if an older binary client sends a request that is evaluated on a newer binary we will not paginate because of CPUOverLimit.

**Review by arulajmani — APPROVED** [review]
:lgtm: 

> Reviewed 9 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-:-NPwmiwg-bgzevjLgeaS:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rhu713, @shermanCRL, and @stevendanna)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwgiNv8oZOnx5rpwCY):*
"erroneously"
___
*[`pkg/kv/kvserver/batcheval/cmd_export.go` line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwlLkUBkkL0DtJNaij:-NPwlLkUBkkL0DtJNaik:b-k61ole) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/kv/kvserver/batcheval/cmd_export.go#L257)):*
> ```Go
> 						// onwards.
> 						if !build.IsRelease() {
> 							return result.Result{}, errors.AssertionFailedf("ExportRequest exited without " +
> ```

Unrelated to your change, in the release build case, should we be logging this? 
___
*[`pkg/kv/kvserver/batcheval/cmd_export.go` line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwlytX7S9M3wefmCMU:-NPwlytX7S9M3wefmCMV:bxb03qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/kv/kvserver/batcheval/cmd_export.go#L313)):*
> ```Go
> 					EndKey: args.EndKey,
> 				}
> 				// TODO(during review): Do we want to add another resume reason
> ```

nit: this looks stale, I assume "RESUME_ELASTIC_CPU_LIMIT" is exactly what this comment is talking about -- should we get rid of this comment? 
___
*[`pkg/sql/sem/builtins/builtins.go` line 7557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwitiz4lSIX6SEi2bF:-NPwitiz4lSIX6SEi2bG:b-oc734x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/sql/sem/builtins/builtins.go#L7557)):*
> ```Go
> 					// DistSender so it likely substantially impacts
> 					// performance.
> 					ReturnElasticCPUResumeSpans: true,
> ```

I may be missing something, but going by the commit message, this function can't handle resume spans. Does that mean we should set this to false?

**Review by stevendanna — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97916)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @arulajmani, @rhu713, and @shermanCRL)

___
*[`-- commits` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwgiNv8oZOnx5rpwCY):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"erroneously"
</blockquote></details>

Thanks! Fixed.
___
*[`pkg/kv/kvpb/api.proto` line 2642 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPXXJ8PFl5EofYKnI87-r1-2642:-NPxLuAME1ECnXZt8rTS:b-a6taqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/kv/kvpb/api.proto#L2642)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: s/unerlying/underlying/g
</blockquote></details>

Thanks! Fixed.
___
*[`pkg/kv/kvserver/batcheval/cmd_export.go` line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwlLkUBkkL0DtJNaij:-NPxLvkoEF9A-mSfZoHK:bnapq4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/kv/kvserver/batcheval/cmd_export.go#L257)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Unrelated to your change, in the release build case, should we be logging this? 
</blockquote></details>

Good idea, now we log a warning.
___
*[`pkg/kv/kvserver/batcheval/cmd_export.go` line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwlytX7S9M3wefmCMU:-NPxM2mV6Ayt8K5tDvxI:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/kv/kvserver/batcheval/cmd_export.go#L313)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: this looks stale, I assume "RESUME_ELASTIC_CPU_LIMIT" is exactly what this comment is talking about -- should we get rid of this comment? 
</blockquote></details>

Removed.
___
*[`pkg/sql/sem/builtins/builtins.go` line 7557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-NPwitiz4lSIX6SEi2bF:-NPxM40nBiU2zcRZlnXX:bmms743) ([raw file](https://github.com/cockroachdb/cockroach/blob/3edc2d546a22fa2dc3c02d55139958dec4b58470/pkg/sql/sem/builtins/builtins.go#L7557)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I may be missing something, but going by the commit message, this function can't handle resume spans. Does that mean we should set this to false? 
</blockquote></details>

I added a bit to the commit message to clarity this.  In this case, the caller knew how to handle the resume spans. But, DistSender needs to know whether a resume span is possible to properly handle the request.  The two lines changed in dist_sender effectively prevents dist_sender from sending the requests in parallel and ensures it expects to get a resume span, avoiding the issue that fingerprinting was hitting.

**Review by stevendanna — COMMENTED** [review]
TFTR! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97916#-:-NPxM_14DrQUDwaKOKmq:bz5m8k4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @arulajmani, @rhu713, and @shermanCRL)

**Comment by stevendanna** [comment]
The lint failure here seems to be a problem inside one of the linters. It passed on a re-run.

bors r=adityamaru,arulajmani

**Pull request merged**